snapd-glib/snapd-notices-monitor.c

```diff
@@ -0,0 +1,185 @@
+            g_autoptr(SnapdNotice) notice = g_object_ref(notices->pdata[i]);
+            g_autoptr(GDateTime) last_occurred = (GDateTime *)snapd_notice_get_last_occurred(notice);
+
+            if (self->last_date_time == NULL || g_date_time_compare(self->last_date_time, last_occurred) <= 0) {
+                g_clear_pointer (&self->last_date_time, g_date_time_unref);
+                self->last_date_time = g_date_time_ref(last_occurred);
+                g_clear_object(&self->last_notice);
+                self->last_notice = g_object_ref(notice);
+            }
+            g_signal_emit_by_name(self, "notice-event", notice);
+        }
+    }
+    begin_monitor(self); // continue waiting for new notifications
+}
+
+static void begin_monitor(SnapdNoticesMonitor *self) {
+    snapd_client_notices_set_after_notice (self->client, self->last_notice);
+    snapd_client_get_notices_async(self->client,
+                                   self->last_date_time,
+                                   2000000000000000, // "infinity" (there is a limit in snapd, around 9 billion seconds)
+                                   self->cancellable,
+                                   (GAsyncReadyCallback)monitor_cb,
+                                   self);
+}
+
+/**
+ * snapd_notices_monitor_start:
+ * @monitor: a #SnapdNoticesMonitor
+ *
+ * Starts the asynchronous listening proccess, that will wait for new
+ * notices and emit a "notice-event" signal with the new notice as
+ * parameter.
+ *
+ * @return: FALSE if there was an error, TRUE if everything worked fine
+ * and the object is listening for events.
+ */
+gboolean snapd_notices_monitor_start(SnapdNoticesMonitor *self, GError **error) {
+
+    g_return_val_if_fail((error == NULL) || (*error == NULL), FALSE);
+    g_return_val_if_fail(SNAPD_IS_NOTICES_MONITOR(self), FALSE);
+    if (self->cancellable != NULL) {
+        *error = g_error_new(SNAPD_ERROR, SNAPD_ERROR_ALREADY_RUNNING, "The notices monitor is already running.");
+        return FALSE;
+    }
+    self->cancellable = g_cancellable_new();
+    begin_monitor(self);
+    return TRUE;
+}
+
+/**
+ * snapd_notices_monitor_stop:
+ * @monitor: a #SnapdNoticesMonitor
+ *
+ * Stops the asynchronous listening proccess started with #snapd_notices_monitor_start.
+ *
+ * @return: FALSE if there was an error, TRUE if everything worked fine.
+ */
+gboolean snapd_notices_monitor_stop(SnapdNoticesMonitor *self, GError **error) {
+
+    g_return_val_if_fail((error == NULL) || (*error == NULL), FALSE);
+    g_return_val_if_fail(SNAPD_IS_NOTICES_MONITOR(self), FALSE);
+    if (self->cancellable == NULL) {
+        *error = g_error_new(SNAPD_ERROR, SNAPD_ERROR_NOT_RUNNING, "The notices monitor isn't running.");
+        return FALSE;
+    }
+    g_cancellable_cancel(self->cancellable);
+    return TRUE;
+}
+
+/**
+ * snapd_notices_monitor_get_client:
+ * @monitor: a #SnapdNoticesMonitor
+ *
+ * Returns the internal #SnapdClient used by this object. Useful if some
```

Comment:
Is this necessary? If additional initialization is required perhaps there should be methods on this object to do those, or the client should be passed into `snapd_notices_monitor_new_with_client()`. Having a getter like this is a risk that a coding mistake could break things.

Reply:
Good point... removing.